File: backend/app/models/gift.py
from datetime import datetime
from sqlalchemy import Column, String, Integer, DateTime, ForeignKey, Boolean, Enum as SQLAlchemyEnum
from sqlalchemy.orm import relationship

from .base import Base
from .types import GiftType

class VirtualGift(Base):
    __tablename__ = "virtual_gifts"

    id = Column(String, primary_key=True)
    name = Column(String, nullable=False)
    description = Column(String)
    image_url = Column(String, nullable=False)
    animation_url = Column(String)
    price = Column(Integer, nullable=False)
    cultural_significance = Column(String)
    cultural_category = Column(String)
    is_premium = Column(Boolean, default=False)

class GiftTransaction(Base):
    __tablename__ = "gift_transactions"

    id = Column(String, primary_key=True)
    gift_id = Column(String, ForeignKey("virtual_gifts.id"))
    sender_id = Column(String, ForeignKey("users.id"))
    receiver_id = Column(String, ForeignKey("users.id"))
    message = Column(String)
    price = Column(Integer, nullable=False)
    created_at = Column(DateTime, default=datetime.utcnow)

class GiftCollection(Base):
    __tablename__ = "gift_collections"

    id = Column(String, primary_key=True)
    user_id = Column(String, ForeignKey("users.id"))
    gift_id = Column(String, ForeignKey("virtual_gifts.id"))
    count = Column(Integer, default=0)
    first_received = Column(DateTime, nullable=False)
    last_received = Column(DateTime, nullable=False)

class Gift(Base):
    __tablename__ = "gifts"

    id = Column(Integer, primary_key=True, index=True)
    from_id = Column(Integer, ForeignKey("users.id", ondelete='CASCADE'), nullable=False)
    to_id = Column(Integer, ForeignKey("users.id", ondelete='CASCADE'), nullable=False)
    gift_type = Column(SQLAlchemyEnum(GiftType), nullable=False)
    created_at = Column(DateTime, default=datetime.utcnow, nullable=False)

    # Relationships
    sender = relationship("User", foreign_keys=[from_id], back_populates="gifts_sent")
    receiver = relationship("User", foreign_keys=[to_id], back_populates="gifts_received")

    def __repr__(self):
        return f"<Gift(from_id={self.from_id}, to_id={self.to_id}, type={self.gift_type})>"

File: pendo_mobile/lib/features/gifts/domain/entities/gift.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'gift.freezed.dart';
part 'gift.g.dart';

@freezed
class VirtualGiftModel with _$VirtualGiftModel {
  const factory VirtualGiftModel({
    required String id,
    required String name,
    required String description,
    @JsonKey(name: 'image_url') required String imageUrl,
    @JsonKey(name: 'animation_url') String? animationUrl,
    required int price,
    @JsonKey(name: 'cultural_significance') String? culturalSignificance,
    @JsonKey(name: 'cultural_category') String? culturalCategory,
    @JsonKey(name: 'is_premium') required bool isPremium,
  }) = _VirtualGiftModel;

  factory VirtualGiftModel.fromJson(Map<String, dynamic> json) =>
      _$VirtualGiftModelFromJson(json);
}

@freezed
class GiftTransactionModel with _$GiftTransactionModel {
  const factory GiftTransactionModel({
    required String id,
    @JsonKey(name: 'gift_id') required String giftId,
    @JsonKey(name: 'sender_id') required String senderId,
    @JsonKey(name: 'receiver_id') required String receiverId,
    String? message,
    required int price,
    @JsonKey(name: 'created_at') required DateTime createdAt,
  }) = _GiftTransactionModel;

  factory GiftTransactionModel.fromJson(Map<String, dynamic> json) =>
      _$GiftTransactionModelFromJson(json);
}

@freezed
class GiftCollectionModel with _$GiftCollectionModel {
  const factory GiftCollectionModel({
    required String id,
    @JsonKey(name: 'user_id') required String userId,
    @JsonKey(name: 'gift_id') required String giftId,
    required int count,
    @JsonKey(name: 'first_received') required DateTime firstReceived,
    @JsonKey(name: 'last_received') required DateTime lastReceived,
  }) = _GiftCollectionModel;

  factory GiftCollectionModel.fromJson(Map<String, dynamic> json) =>
      _$GiftCollectionModelFromJson(json);
}
